addons/binding/org.openhab.binding.homepilot/src/main/java/org/openhab/binding/homepilot/discovery/HomePilotDeviceDiscoveryService.java

```diff
@@ -0,0 +1,162 @@
+ *
+ * @author Gerhard Riegler - Initial contribution
+ */
+public class HomePilotDeviceDiscoveryService extends AbstractDiscoveryService implements ExtendedDiscoveryService {
+    private static final Logger logger = LoggerFactory.getLogger(HomePilotDeviceDiscoveryService.class);
+    private static final int DISCOVER_TIMEOUT_SECONDS = 300;
+
+    private HomePilotBridgeHandler bridgeHandler;
+    private Future<?> scanFuture;
+    private DiscoveryServiceCallback discoveryServiceCallback;
+
+    public HomePilotDeviceDiscoveryService(HomePilotBridgeHandler bridgeHandler) {
+        super(ImmutableSet.of(new ThingTypeUID(BINDING_ID, "-")), DISCOVER_TIMEOUT_SECONDS, false);
+        this.bridgeHandler = bridgeHandler;
+    }
+
+    /**
+     * Called on component activation.
+     */
+    public void activate() {
+        super.activate(null);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void deactivate() {
+        super.deactivate();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void startScan() {
+        logger.info("Starting HomePilot discovery scan");
+        loadDevices();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void stopScan() {
+        logger.info("Stopping HomePilot discovery scan");
+        bridgeHandler.getGateway().cancelLoadAllDevices();
+        waitForScanFinishing();
+        super.stopScan();
+    }
+
+    /**
+     * Waits for the discovery scan to finish and then returns.
+     */
+    public void waitForScanFinishing() {
+        if (scanFuture != null) {
+            logger.info("Waiting for finishing HomePilot device discovery scan");
+            try {
+                scanFuture.get();
+                logger.debug("HomePilot device discovery scan finished");
+            } catch (CancellationException ex) {
+                // ignore
+            } catch (Exception ex) {
+                logger.error("Error waiting for device discovery scan: {}", ex.getMessage(), ex);
+            }
+        }
+    }
+
+    /**
+     * Starts a thread which loads all Homematic devices connected to the gateway.
+     */
+    public void loadDevices() {
+        if (scanFuture == null) {
+            scanFuture = scheduler.submit(new Runnable() {
+
+                @Override
+                public void run() {
+                    try {
+                        List<HomePilotDevice> allKnownDevices = bridgeHandler.getGateway().loadAllDevices();
+                        // discover all things
+                        for (HomePilotDevice device : allKnownDevices) {
+                            ThingUID uid = createUID(device);
+                            deviceDiscovered(uid, device);
+                        }
+
+                        logger.debug("Finished HomePilot device discovery scan on gateway '{}'",
+                                bridgeHandler.getGateway().getId());
+                    } catch (Throwable ex) {
```

Comment:
Do not catch `Throwable`'s you will also catch for example `OutOfMemoryError`s, nothing you can do about that, logging an error will not resolve it ;-) 
In this case catching an `RuntimeException` might be legitimite but in general that is also a bad practice.